aspnetcore/web-api/error-handling.md

```diff
@@ -0,0 +1,201 @@
+
+```csharp
+public class HttpResponseException : Exception
+{
+    public int Status { get; set; } = 500;
+
+    public object Value { get; set; }
+}
+```
+
+```csharp
+public class HttpResponseExceptionFilter : IActionFilter, IOrderedFilter
+{
+    public int Order { get; set; } = int.MaxValue - 10;
+
+    public void OnActionExecuting(ActionExecutingContext context) {}
+
+    public void OnActionExecuted(ActionExecutedContext context)
+    {
+        if (context.Exception is HttpResponseException exception)
+        {
+            context.Result = new ObjectResult(exception.Value)
+            {
+                Status = exception.Status,
+            };
+            context.ExceptionHandled = true;
+        }
+    }
+}
+```
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddControllers(options => 
+        options.Filters.Add(new HttpResponseExceptionFilter()));
+}
+```
+
+## Validation failure error response
+
+For web API controllers, MVC responds with a `ValidationProblemDetails` response type when model validation fails. MVC uses the results of <xref:Microsoft.AspNetCore.Mvc.ApiBehaviorOptions.InvalidModelStateResponseFactory> to construct
+the error response for a validation failure. The following example uses the factory to change the default response type to `SerializableError`:
+
+::: moniker range=">= aspnetcore-3.0"
+
+[!code-csharp[](index/samples/3.x/Startup.cs?name=snippet_DisableProblemDetailsInvalidModelStateResponseFactory&highlight=4-13)]
+
+::: moniker-end
+
+::: moniker range="= aspnetcore-2.2"
+
+[!code-csharp[](index/samples/2.x/Startup.cs?name=snippet_DisableProblemDetailsInvalidModelStateResponseFactory&highlight=5-14)]
+
+::: moniker-end
+
+::: moniker range=">= aspnetcore-3.0"
+
+## Client error response
+
+For web API controllers, MVC transforms an error result (a result with status code 400 or higher) to a result with <xref:Microsoft.AspNetCore.Mvc.ProblemDetails>. The error response can be configured in one of the following ways:
+
+1. [Use ApiBehaviorOptions.ClientErrorMapping](#use-apibehavioroptionsclienterrormapping)
+1. [Implement ProblemDetailsFactory](#implement-problemdetailsfactory)
+
+### Use ApiBehaviorOptions.ClientErrorMapping
+
+Use the <xref:Microsoft.AspNetCore.Mvc.ApiBehaviorOptions.ClientErrorMapping*> property to configure the contents of the `ProblemDetails` response. For example, the following code updates the `type` property for 404 responses:
+
+::: moniker range=">= aspnetcore-3.0"
+
+[!code-csharp[](index/samples/3.x/Startup.cs?name=snippet_ConfigureApiBehaviorOptions&highlight=8-9)]
+
+::: moniker-end
+
+::: moniker range="<= aspnetcore-2.2"
+
+[!code-csharp[](index/samples/2.x/Startup.cs?name=snippet_ConfigureApiBehaviorOptions&highlight=9-10)]
+
+::: moniker-end
+
+### Implement ProblemDetailsFactory
```

Comment:
@pranavkm Is `ProblemDetailsFactory` available only in 3.0?

Reply:
Yup. It's new in 3.0